File: 539_MinimumTimeDifference.py
##54.3%
class Solution(object):
    def findMinDifference(self, timePoints):
        """
        :type timePoints: List[str]
        :rtype: int
        """
        timePoints.sort()
        timeList = map(lambda timePoint: int(timePoint[:2]) * 60 + int(timePoint[3:]), timePoints)
        timeList.append(timeList[0] + 24 * 60)
        minDiff = 0x7fffffff
        for prevTime, currTime in zip(timeList[:len(timeList) - 1], timeList[1: len(timeList)]):
            minDiff = min(minDiff, currTime - prevTime)
        return minDiff
		
class Solution(object):
    def findMinDifference(self, timePoints):
        """
        :type timePoints: List[str]
        :rtype: int
        """
        timePoints.sort()
        timeList = map(lambda timePoint: (int(timePoint[:2]), int(timePoint[3:])), timePoints)
        timeList.append((timeList[0][0] + 24, timeList[0][1]))
        minDiff = (0x7fffffff, 0x7fffffff)
        for prevTime, currTime in zip(timeList[:len(timeList) - 1], timeList[1: len(timeList)]):
            minDiff = min(minDiff, (currTime[0] - prevTime[0], currTime[1] - prevTime[1]))
        return minDiff[0] * 60 + minDiff[1]
		
def findMinDifference(self, A):
    def convert(time):
        return int(time[:2]) * 60 + int(time[3:])
    minutes = map(convert, A)
    minutes.sort()
    
    return min( (y - x) % (24 * 60) 
                for x, y in zip(minutes, minutes[1:] + minutes[:1]) )

File: 539_MinimumTimeDifference.cpp
//60.93%
class Solution {
public:
    int findMinDifference(vector<string>& timePoints) {
        sort(timePoints.begin(), timePoints.end());
		int minDiff(INT_MAX);
		int preHour((timePoints[0][0] - '0') * 10 + timePoints[0][1] - '0');
		int preMinute((timePoints[0][3] - '0') * 10 + timePoints[0][4] - '0');
		int firstHour(preHour), firstMinute(preMinute);
		for(int i = 1; i < timePoints.size(); ++i){
			int currHour((timePoints[i][0] - '0') * 10 + timePoints[i][1] - '0');
			int currMinute((timePoints[i][3] - '0') * 10 + timePoints[i][4] - '0');
			minDiff = min(minDiff, (currHour - preHour) * 60 + (currMinute - preMinute));
			preHour = currHour;
			preMinute = currMinute;
		}
		minDiff = min(minDiff, (24 + firstHour - preHour) * 60 + firstMinute - preMinute);
		return minDiff;
    }
};

//There is only 24 * 60 = 1440 possible time points. Just create a boolean array, each element stands for if we see that time point or not. Then things become simple...

public class Solution {
    public int findMinDifference(List<String> timePoints) {
        boolean[] mark = new boolean[24 * 60];
        for (String time : timePoints) {
            String[] t = time.split(":");
            int h = Integer.parseInt(t[0]);
            int m = Integer.parseInt(t[1]);
            if (mark[h * 60 + m]) return 0;
            mark[h * 60 + m] = true;
        }
        
        int prev = 0, min = Integer.MAX_VALUE;
        int first = Integer.MAX_VALUE, last = Integer.MIN_VALUE;
        for (int i = 0; i < 24 * 60; i++) {
            if (mark[i]) {
                if (first != Integer.MAX_VALUE) {
                    min = Math.min(min, i - prev);
                }
                first = Math.min(first, i);
                last = Math.max(last, i);
                prev = i;
            }
        }
        
        min = Math.min(min, (24 * 60 - last + first));
        
        return min;
    }
}

//Similar like mine
class Solution {
public:
    int findMinDifference(vector<string>& times) {
        int n = times.size();
        sort(times.begin(), times.end());
        int mindiff = INT_MAX;
        for (int i = 0; i < times.size(); i++) {
            int diff = abs(timeDiff(times[(i - 1 + n) % n], times[i]));
            diff = min(diff, 1440 - diff);
            mindiff = min(mindiff, diff);
        }
        return mindiff;
    }

private:
    int timeDiff(string t1, string t2) {
        int h1 = stoi(t1.substr(0, 2));
        int m1 = stoi(t1.substr(3, 2));
        int h2 = stoi(t2.substr(0, 2));
        int m2 = stoi(t2.substr(3, 2));
        return (h2 - h1) * 60 + (m2 - m1);
    }
};

public class Solution {
    public int findMinDifference(List<String> timePoints) {
        int mm = Integer.MAX_VALUE;
        List<Integer> time = new ArrayList<>();
        
        for(int i = 0; i < timePoints.size(); i++){
            Integer h = Integer.valueOf(timePoints.get(i).substring(0, 2));
            time.add(60 * h + Integer.valueOf(timePoints.get(i).substring(3, 5)));
        }
        
        Collections.sort(time, (Integer a, Integer b) -> a - b);
        
        for(int i = 1; i < time.size(); i++){
            System.out.println(time.get(i));
            mm = Math.min(mm, time.get(i) - time.get(i-1));
        }
        
        int corner = time.get(0) + (1440 - time.get(time.size()-1));
        return Math.min(mm, corner);
    }
}